engine/runtime/src/main/scala/org/enso/compiler/pass/resolve/FullyQualifiedNames.scala

```diff
@@ -0,0 +1,336 @@
+              lit
+          }
+        } else {
+          lit
+        }
+      case app @ IR.Application.Prefix(_, List(_), _, _, _, _) =>
+        app.function match {
+          case lit: IR.Name.Literal =>
+            if (lit.isMethod)
+              resolveLocalApplication(
+                app,
+                bindings,
+                freshNameSupply,
+                compiler,
+                selfTypeResolution
+              )
+            else
+              app.mapExpressions(
+                processExpression(
+                  _,
+                  bindings,
+                  freshNameSupply,
+                  selfTypeResolution,
+                  compiler
+                )
+              )
+          case _ =>
+            app.mapExpressions(
+              processExpression(
+                _,
+                bindings,
+                freshNameSupply,
+                selfTypeResolution,
+                compiler
+              )
+            )
+
+        }
+
+    }
+
+  private def resolveLocalApplication(
+    app: IR.Application.Prefix,
+    bindings: BindingsMap,
+    freshNameSupply: FreshNameSupply,
+    compiler: Option[Compiler],
+    selfTypeResolution: Option[Resolution]
+  ): IR.Expression = {
+    val processedFun =
+      processExpression(
+        app.function,
+        bindings,
+        freshNameSupply,
+        selfTypeResolution,
+        compiler
+      )
+    val processedArgs =
+      app.arguments.map(
+        _.mapExpressions(
+          processExpression(
+            _,
+            bindings,
+            freshNameSupply,
+            selfTypeResolution,
+            compiler
+          )
+        )
+      )
+
+    val processedApp = processedArgs match {
+      case List(thisArg) =>
+        (thisArg.value.getMetadata(this).map(_.target), processedFun) match {
+          case (Some(resolved @ ResolvedLibrary(_)), name: IR.Name.Literal) =>
+            resolveQualName(resolved, name, compiler).map(resolvedMod =>
+              freshNameSupply
+                .newName()
+                .updateMetadata(this -->> resolvedMod)
+                .setLocation(name.location)
+            )
+          case _ =>
+            None
+        }
+      case _ =>
+        None
+    }
+
+    processedApp.getOrElse(
+      app.copy(function = processedFun, arguments = processedArgs)
+    )
+  }
+
+  private def resolveQualName(
+    thisResolution: ResolvedLibrary,
+    consName: IR.Name.Literal,
+    optCompiler: Option[Compiler]
+  ): Option[FQNResolution] = {
+    optCompiler.flatMap { compiler =>
+      val pkgRepository = compiler.packageRepository
+      val libName       = LibraryName(thisResolution.namespace, consName.name)
+      pkgRepository.ensurePackageIsLoaded(libName).toOption.flatMap { _ =>
+        pkgRepository
+          .getLoadedModule(
+            s"${libName.toString}.${Imports.mainModuleName.name}"
+          )
+          .map { m =>
+            if (m.getIr == null) {
```

Comment:
This change is probably the most controversial as it means triggering a full compilation of the library.

We could limit the scope of this change and issue an error here instead. In that case, we wouldn't be able to use a fully qualified names unless someone imported (parts of) it before. So far, that would be extremely rare since regular programs are rather limited without stdlib.

Reply:
I like limiting FQNs to just imported libraries. I'm not sure if that's what we want as at this point they are less useful (wasn't their whole point the fact that you can use them without imports?).

But I think, especially as in Enso the imports actually _define_ which packages a project relies on, it's good to have one place where this is define - i.e. the imports which are relatively easy to glance. If we add FQNs within the code, it will be much harder to find out which libraries are _actually_ imported.

So I would vote for indeed only allowing to use an FQN on an imported library, but I understand that we may need to do that differently due to other requirements, so I'm also not opposed to allowing FQNs to load libraries if that is needed.

Reply:
That can be reduced fairly easy by reporting the error in the condition rather than triggering `compiler.run`.
That would mean however that
```
main =
  Standard.Base.IO.println "Hello world!"
```
needs to be
```
import Standard.Base
main =
  Standard.Base.IO.println "Hello world!"
```
at least.

Reply:
I like this, we can make the error quite readable ("FQN used, but its library is not imported, please add the missing `import Foo.Bar` import" - we can even include the code in the error message).
The IDE should also be able to handle adding the proper import. It will also be easier for the IDE too to track stuff if all library dependencies are only in the imports and not scattered around the whole codebase.

Reply:
But I'm not the one who was requesting making FQN's work, so probably needs their opinion too to make sure such behaviour would be ok. I'm mostly looking from the perspective of making it easy to track the dependencies.

Reply:
```py
main =
  Standard.Base.IO.println "Hello world!"
```
I think we can schedule compilation of `Stdandard.Base.IO`, and the imports resolver algorithm will figure out all the required dependencies. Although, it can be implemented as an optimization in the future. I feel it's ok to trigger the compilation of `Standard.Base` in this case.

Reply:
The IDE currently `from Standard.Base import all`. Then the whole _standard base library_ should be compiled and one could use FQN to refer to any of type, right? If so, then we may not need to cyclic dependency from `ModuleContext` to `Compiler` probably. But I am not a language designer.

Reply:
So I feel like it would be safer to limit this PR on purpose as there are a lot of unknowns here. We could come back to this if there is really a demand for import-less FQNs.

Reply:
After all, I think the IDE is supposed to have a whole import-management capability, so this seems reasonable, especially if we report a friendly error

Reply:
done